### PR TITLE
Shipping Method Settings: Update modal contents

### PR DESCRIPTION
**Pull request by psealock**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/40189

Update modal contents for Shipping Zone Methods and Shipping Classes according to designs found in paJDVv-2Ew-p2.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Head to `/wp-admin/admin.php?page=wc-settings&tab=shipping` on a Jurassic Ninja site with this branch
2. Add a Shipping Zone with Name and Regions added.
3. Click "Add shipping method" and go through the flow for Flat Rate, Free Shipping and Local Pickup the modals should look like the designs.
5. Be sure to select a few options for `FREE SHIPPING REQUIRES` to see the rest of the form change dynamically.
6. Now add a Shipping Class from `/wp-admin/admin.php?page=wc-settings&tab=shipping&section=classes`. That modal should also reflect the new designs.
7. Now that you've added a Shipping Zone Class, return to the Shipping Zone Methods screen. Add or edit a Flat Rate method, you should now see the configuration modal include fields pertinent to your Shipping Class.
8. Now add a new method from an extension. [Flat Rate Boxes](https://github.com/woocommerce/woocommerce-shipping-flat-rate-boxes/releases), for example. Add a new method Flat Rate Boxes. Note, it doesn't have a configuration screen, so the second modal is omitted.

### Known issues

The following are known and will be addressed in a follow up.

Between Adding and configuring a Shipping Method, the button's spinner needs help
image

After a Shipping Class is added, configuring a Flat Rate method includes new fields. Since we programatically append the help text underneath the input, these look off since they are so long. @FlorianBru should we go back to the help tooltip for these or simply change the text?

image

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Update Shipping Class modals to new style

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js [hunk 1]
```diff
@@ -461,12 +540,49 @@
 						}, 'json' );
 					}
 				},
+				// Free Shipping has hidden field elements depending on data values.
+				possiblyHideFreeShippingRequirements: function( data ) {
+					if ( Object.keys( data ).includes( 'woocommerce_free_shipping_requires' ) ) {
+						const shouldHideRequirements = data.woocommerce_free_shipping_requires === '' || data.woocommerce_free_shipping_requires === 'coupon';
```
Comment (ilyasfoo) [inline plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js]: Could `data.woocommerce_free_shipping_requires` be `null` here?

File: plugins/woocommerce/includes/shipping/free-shipping/class-wc-shipping-free-shipping.php [hunk 2]
```diff
@@ -89,19 +89,19 @@ public function init() {
...
+				'title'   => __( 'Free shipping requires', 'woocommerce' ),
 				'type'    => 'select',
 				'class'   => 'wc-enhanced-select',
 				'default' => '',
 				'options' => array(
-					''           => __( 'N/A', 'woocommerce' ),
+					''           => __( 'No requirement', 'woocommerce' ),
 					'coupon'     => __( 'A valid free shipping coupon', 'woocommerce' ),
 					'min_amount' => __( 'A minimum order amount', 'woocommerce' ),
 					'either'     => __( 'A minimum order amount OR a coupon', 'woocommerce' ),
```
Comment (ilyasfoo) [inline plugins/woocommerce/includes/shipping/free-shipping/class-wc-shipping-free-shipping.php]: Not related to this PR but just pointing out in the design it's `A minimum order amount OR coupon`.

**Review by ilyasfoo — COMMENTED** [review]
@psealock Functionality wise this is working well! However, there's many small differences between the UI and the design - I'm not sure if I'm supposed to list them all. Let me know if I should!

image

The shipping class costs link is not implemented.

image

1. The ordering of shipping method is different than design.

File: plugins/woocommerce/includes/shipping/free-shipping/class-wc-shipping-free-shipping.php (hunk 2 above)
Comment (psealock) [reply plugins/woocommerce/includes/shipping/free-shipping/class-wc-shipping-free-shipping.php]: Fixed in https://github.com/woocommerce/woocommerce/pull/40615/commits/78ecd994be3cf1611b6a5a0e50f428918035d963

File: plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js (hunk 1 above)
Comment (psealock) [reply plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js]: Probably not, but good idea to check for null as well. Fixed in https://github.com/woocommerce/woocommerce/pull/40615/commits/afdef12962e5e34e8f4d9569b30262de59d3425e

**Comment by psealock** [comment]
Thanks for the review @ilyasfoo !

>However, there's many small differences between the UI and the design - I'm not sure if I'm supposed to list them all. Let me know if I should!

Yes please! Indeed there are a few, thanks for mentioning a some. I handled a few and created issues for others to follow up on.

Changes made:
* Fixed the order of Shipping Zone Methods https://github.com/woocommerce/woocommerce/pull/40615/commits/f71a7c3e90c52f65fdd67a42111fa102b974b20c
* Fixed the primary action button on modals to reflect the designs https://github.com/woocommerce/woocommerce/pull/40615/commits/a321c972dd40ec4f66985990774bf3a7406ebc67

Created new issues for follow ups
* https://github.com/woocommerce/woocommerce/issues/40736
* https://github.com/woocommerce/woocommerce/issues/40737
* https://github.com/woocommerce/woocommerce/issues/40738

**Review by ilyasfoo — COMMENTED** [review]
@psealock Changes look good! Here are my comments on the UI, but I don't think everything is mandatory.

Minor differences in create shipping modal from design ([screenshot](https://github.com/woocommerce/woocommerce/assets/3747241/65c72cd8-9948-4b6a-a14e-792447919dfc)): 
1. Insufficient vertical spacing
2. Insufficient vertical spacing
3. Different icon used
4. Different icon used
5. Different icon used
6. Insufficient top spacing
7. Different color or weight 

Minor differences in configure modal ([screenshot](https://github.com/woocommerce/woocommerce/assets/3747241/cc95fb41-2ac7-4910-892b-8118bc6ad392)):

1. Insufficient bottom spacing
2. Too much vertical spacing
3. Insufficient vertical spacing
4. Misalignment between checkbox and label, checkbox too close with label
5. Button is enabled even when shipping name is not set

Design isn't available for a custom shipping class, but here's my feedback ([screenshot](https://github.com/woocommerce/woocommerce/assets/3747241/4da8e2fe-a70e-4022-9934-97e41f472243))

1. Insufficient vertical spacing
2. Insufficient vertical spacing

The button color and active border does not follow theme color ([screenshot](https://github.com/woocommerce/woocommerce/assets/3747241/7d275f6a-d4ce-428a-a981-6f4934cdc2bd))
 
Select component is not used in dropdown ([screenshot](https://github.com/woocommerce/woocommerce/assets/3747241/5e7379d2-4e6a-42d4-94cb-2fd5c66a9c38))

**Comment by psealock** [comment]
Thanks for another great review @ilyasfoo  - I'm getting the hang of reading the Figma files now, so these changes should look much better.

Screenshot 2023-10-17 at 12 31 10 PM

Screenshot 2023-10-17 at 12 31 24 PM

Screenshot 2023-10-17 at 12 31 44 PM

>Different icon used

I need to stick with the available dash-icons because this isn't in the new Gutenberg context

>Select component is not used in dropdown

Same with this. I need to stick with whats available

>Button is enabled even when shipping name is not set

I added some minor validation for the configure screen just on the `name` field. Since this is extendable, I will revisit this validation in a follow up.

>The button color and active border does not follow theme color

Great call out here, I didn't even think about this. Since this involves more than buttons I will tackle this in a follow up PR.

**Comment by psealock** [comment]
I'm going to go ahead and merge this into the feature branch in order to keep things moving and since changes requested were visual, I can address them in follow ups.

**Pull request merged**